File: packages/python/goldencheck-types/goldencheck_types/loader.py
"""Load domain packs from yaml files."""
from __future__ import annotations

import functools
import os
from pathlib import Path

import yaml

from goldencheck_types.types import DomainPack, FieldSpec


class DomainPackError(ValueError):
    """A domain-pack YAML file is malformed (wrong shape, type, or value).

    Distinct from FileNotFoundError (file missing) and KeyError (unknown
    domain name) so callers can react differently — a malformed pack is
    a fix-the-yaml situation, not a fix-the-call situation.
    """


def _domains_dir() -> Path:
    """Resolve the domains/ directory at runtime.

    Order:
    1. Test override via ``GOLDENCHECK_TYPES_TEST_DIR`` env var.
    2. Vendored at ``goldencheck_types/_domains/`` — present both in
       source checkouts and in built wheels / sdists. This is the
       authoritative location for the Python package.
    3. Cross-package monorepo fallback:
       ``packages/typescript/goldencheck-types/domains/``. Only used
       when the vendored copy is absent (e.g. a fresh source checkout
       before ``scripts/sync-domain-packs.py`` has run). Going through
       this path means the YAMLs are NOT in the wheel — every install
       outside the monorepo would break — so the vendored dir should
       always be preferred.
    """
    if override := os.environ.get("GOLDENCHECK_TYPES_TEST_DIR"):
        return Path(override)

    here = Path(__file__).resolve().parent

    bundled = here / "_domains"
    if bundled.exists() and any(bundled.glob("*.yaml")):
        return bundled

    # Source-checkout fallback — useful only inside the monorepo before
    # the vendoring step has run. Production installs always hit the
    # bundled branch above.
    source_layout = (
        here.parent.parent.parent
        / "typescript"
        / "goldencheck-types"
        / "domains"
    )
    if source_layout.exists():
        return source_layout

    raise FileNotFoundError(f"Could not locate domains/ near {here}")


def list_domains() -> list[str]:
    return sorted(p.stem for p in _domains_dir().glob("*.yaml"))


def clear_cache() -> None:
    """Drop memoized domain packs.

    Tests that mutate YAML files on disk after the first ``load_domain``
    call must invoke this — otherwise subsequent loads return the cached
    pre-mutation pack. Production code never needs this; ``load_domain``
    keys its cache on the resolved domains directory, so flipping
    ``GOLDENCHECK_TYPES_TEST_DIR`` between calls invalidates naturally.
    """
    _load_domain_cached.cache_clear()


@functools.lru_cache(maxsize=32)
def _load_domain_cached(name: str, domains_dir: str) -> DomainPack:
    return _load_domain_uncached(name, Path(domains_dir))


def load_domain(name: str) -> DomainPack:
    """Load and validate a domain pack YAML (memoized; see ``clear_cache``).

    Shape-checks every field rather than silently coercing. A misindented
    ``name_hints:`` or a string-where-list-expected used to produce a pack
    that "loaded fine" but matched nothing (or matched everything via
    single-character iteration over a string); now it raises
    ``DomainPackError`` with the file path and key path so the user can
    fix the YAML directly.
    """
    return _load_domain_cached(name, str(_domains_dir()))


def _load_domain_uncached(name: str, domains_dir: Path) -> DomainPack:
    path = domains_dir / f"{name}.yaml"
    if not path.exists():
        raise KeyError(f"domain pack {name!r} not found in {domains_dir}")

    raw = yaml.safe_load(path.read_text(encoding="utf-8"))
    if raw is None:
        raise DomainPackError(f"{path}: empty or null YAML; expected a mapping")
    if not isinstance(raw, dict):
        raise DomainPackError(
            f"{path}: top level must be a mapping, got {type(raw).__name__}"
        )

    types_raw = raw.get("types")
    if types_raw is None:
        types_raw = {}
    elif not isinstance(types_raw, dict):
        raise DomainPackError(
            f"{path}: 'types' must be a mapping, got {type(types_raw).__name__}"
        )

    types: dict[str, FieldSpec] = {}
    for type_name, spec in types_raw.items():
        if not isinstance(spec, dict):
            raise DomainPackError(
                f"{path}: types.{type_name} must be a mapping, got {type(spec).__name__}"
            )

        name_hints = spec.get("name_hints", [])
        if not isinstance(name_hints, list):
            raise DomainPackError(
                f"{path}: types.{type_name}.name_hints must be a list, "
                f"got {type(name_hints).__name__}"
            )

        value_signals = spec.get("value_signals", {})
        if not isinstance(value_signals, dict):
            raise DomainPackError(
                f"{path}: types.{type_name}.value_signals must be a mapping, "
                f"got {type(value_signals).__name__}"
            )

        suppress = spec.get("suppress", [])
        if not isinstance(suppress, list):
            raise DomainPackError(
                f"{path}: types.{type_name}.suppress must be a list, "
                f"got {type(suppress).__name__}"
            )

        threshold = spec.get("confidence_threshold")
        if threshold is not None:
            try:
                threshold_f = float(threshold)
            except (TypeError, ValueError):
                raise DomainPackError(
                    f"{path}: types.{type_name}.confidence_threshold must be numeric, "
                    f"got {threshold!r}"
                )
            if not (0.0 <= threshold_f <= 1.0):
                raise DomainPackError(
                    f"{path}: types.{type_name}.confidence_threshold must be in [0,1], "
                    f"got {threshold!r}"
                )
        else:
            threshold_f = None

        # If the YAML explicitly carries a `name:` it must match the dict
        # key. Disagreement signals user error (typo, copy-paste). The
        # YAMLs ship without a `name:` today; loader populates from the
        # key.
        explicit_name = spec.get("name")
        if explicit_name is not None and explicit_name != type_name:
            raise DomainPackError(
                f"{path}: types.{type_name}.name is {explicit_name!r}, "
                f"but it lives under key {type_name!r}. The two must agree.",
            )
        types[type_name] = FieldSpec(
            name=type_name,
            name_hints=[str(h) for h in name_hints],
            value_signals=dict(value_signals),
            suppress=[str(s) for s in suppress],
            confidence_threshold=threshold_f,
            description=spec.get("description"),
        )

    return DomainPack(
        name=name,
        description=raw.get("description") or "",
        types=types,
    )

File: packages/typescript/goldencheck-types/src/loader.ts
// Load domain packs from yaml files at runtime.
import * as fs from "fs";
import * as path from "path";
import * as url from "url";
import * as yaml from "js-yaml";
import type { DomainPack, FieldSpec } from "./types.js";

/** A domain-pack YAML file is malformed (wrong shape, type, or value).
 *  Distinct from "file not found" so callers can react differently — a
 *  malformed pack is a fix-the-yaml situation, not a fix-the-call situation. */
export class DomainPackError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "DomainPackError";
  }
}

function isPlainObject(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

function domainsDir(): string {
  if (process.env.GOLDENCHECK_TYPES_TEST_DIR) {
    return process.env.GOLDENCHECK_TYPES_TEST_DIR;
  }
  // Resolve relative to this file. In dev (src/) we go up one to find domains/;
  // in dist/ we also go up one (dist/loader.js -> ../domains).
  const here = path.dirname(url.fileURLToPath(import.meta.url));
  return path.resolve(here, "..", "domains");
}

// Module-scoped pack cache. Keyed on (resolved-domains-dir, pack-name) so
// flipping GOLDENCHECK_TYPES_TEST_DIR between calls invalidates naturally
// without callers remembering to clearCache().
const _packCache = new Map<string, DomainPack>();

/** Drop memoized domain packs.
 *
 *  Tests that mutate YAML files on disk after the first `loadDomain` call
 *  must invoke this — otherwise subsequent loads return the cached
 *  pre-mutation pack. Production code never needs this. */
export function clearCache(): void {
  _packCache.clear();
}

export function listDomains(): string[] {
  return fs
    .readdirSync(domainsDir())
    .filter((f) => f.endsWith(".yaml"))
    .map((f) => f.replace(/\.yaml$/, ""))
    .sort();
}

/** Load and shape-validate a domain pack YAML (memoized; see `clearCache`).
 *
 * Mismatched types raise `DomainPackError` with the file path + key path
 * so the user can fix the YAML directly. Previously a misindented
 * `name_hints:` or a string-where-list-expected produced a pack that
 * "loaded fine" but matched nothing at runtime.
 */
export function loadDomain(name: string): DomainPack {
  const dir = domainsDir();
  const cacheKey = `${dir}::${name}`;
  const cached = _packCache.get(cacheKey);
  if (cached) return cached;

  const filePath = path.join(dir, `${name}.yaml`);
  if (!fs.existsSync(filePath)) {
    throw new Error(
      `domain pack '${name}' not found in ${dir}`,
    );
  }
  const parsed = yaml.load(fs.readFileSync(filePath, "utf-8")) as unknown;
  if (parsed === null || parsed === undefined) {
    throw new DomainPackError(
      `${filePath}: empty or null YAML; expected a mapping`,
    );
  }
  if (!isPlainObject(parsed)) {
    throw new DomainPackError(
      `${filePath}: top level must be a mapping, got ${typeof parsed}`,
    );
  }
  const raw = parsed;
  const rawTypesAny = raw.types;
  let rawTypes: Record<string, unknown>;
  if (rawTypesAny === undefined || rawTypesAny === null) {
    rawTypes = {};
  } else if (!isPlainObject(rawTypesAny)) {
    throw new DomainPackError(
      `${filePath}: 'types' must be a mapping, got ${typeof rawTypesAny}`,
    );
  } else {
    rawTypes = rawTypesAny;
  }

  const types: Record<string, FieldSpec> = {};
  for (const [typeName, specAny] of Object.entries(rawTypes)) {
    if (!isPlainObject(specAny)) {
      throw new DomainPackError(
        `${filePath}: types.${typeName} must be a mapping, got ${typeof specAny}`,
      );
    }
    const spec = specAny;

    const nameHints = spec.name_hints ?? [];
    if (!Array.isArray(nameHints)) {
      throw new DomainPackError(
        `${filePath}: types.${typeName}.name_hints must be a list, got ${typeof nameHints}`,
      );
    }
    const valueSignals = spec.value_signals ?? {};
    if (!isPlainObject(valueSignals)) {
      throw new DomainPackError(
        `${filePath}: types.${typeName}.value_signals must be a mapping, got ${typeof valueSignals}`,
      );
    }
    const suppress = spec.suppress ?? [];
    if (!Array.isArray(suppress)) {
      throw new DomainPackError(
        `${filePath}: types.${typeName}.suppress must be a list, got ${typeof suppress}`,
      );
    }
    const threshold = spec.confidence_threshold as number | undefined;
    if (threshold !== undefined) {
      if (typeof threshold !== "number" || Number.isNaN(threshold)) {
        throw new DomainPackError(
          `${filePath}: types.${typeName}.confidence_threshold must be numeric, got ${threshold}`,
        );
      }
      if (threshold < 0 || threshold > 1) {
        throw new DomainPackError(
          `${filePath}: types.${typeName}.confidence_threshold must be in [0,1], got ${threshold}`,
        );
      }
    }

    // If the YAML explicitly carries a `name:` it must match the dict
    // key. Disagreement signals user error (typo, copy-paste).
    if (
      spec.name !== undefined &&
      spec.name !== null &&
      spec.name !== typeName
    ) {
      throw new DomainPackError(
        `${filePath}: types.${typeName}.name is ${JSON.stringify(spec.name)}, ` +
          `but it lives under key ${JSON.stringify(typeName)}. The two must agree.`,
      );
    }
    types[typeName] = {
      name: typeName,
      name_hints: nameHints.map(String),
      value_signals: valueSignals,
      suppress: suppress.map(String),
      ...(threshold !== undefined ? { confidence_threshold: threshold } : {}),
      ...(typeof spec.description === "string" ? { description: spec.description } : {}),
    };
  }
  const pack: DomainPack = {
    name,
    description: typeof raw.description === "string" ? raw.description : "",
    types,
  };
  _packCache.set(cacheKey, pack);
  return pack;
}
